{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92c843c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use standard multiprocessing data transfer (pipe) to transfer data between the main process and workers.\n",
      "\n",
      "WARNING: You are on Windows. If you detect any issue with pandarallel, be sure you checked out the Troubleshooting page:\n",
      "https://nalepae.github.io/pandarallel/troubleshooting/\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import itertools as it\n",
    "from pandarallel import pandarallel\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.model_selection import RepeatedKFold\n",
    "\n",
    "import random\n",
    "\n",
    "from sklearn.linear_model import LinearRegression\n",
    "pandarallel.initialize()\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "618252b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_input_path = lambda fname: os.path.abspath(\"/home/people/18200264/local_data/input_data/\"+ fname)\n",
    "# get_output_path = lambda fname: os.path.abspath(\"/home/people/18200264/local_data/output/GuanLab/CPTAC/\"+ fname)\n",
    "\n",
    "get_input_path = lambda fname: os.path.normpath('../local_data/processed_data/'+ fname)\n",
    "get_output_path = lambda fname: os.path.normpath('../local_data/results/CPTAC/'+ fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cefeb196",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_cptac_transcriptomics = get_input_path('CPTAC_Transcriptomics_processed.parquet')\n",
    "file_cptac_proteomics = get_input_path('CPTAC_Proteomics_processed.parquet')\n",
    "file_cptac_sample_info = get_input_path('CPTAC_sample_info.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f652d35d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dimensions:  (2021, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00004</th>\n",
       "      <td>ccRCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00010</th>\n",
       "      <td>ccRCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Study\n",
       "C3L-00004  ccRCC\n",
       "C3L-00010  ccRCC"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_sample = pd.read_parquet(file_cptac_sample_info)[['Study']]\n",
    "print(\"Dimensions: \", cptac_sample.shape)\n",
    "cptac_sample[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6e0ebd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14792, 1227)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>01OV007</th>\n",
       "      <th>01OV017</th>\n",
       "      <th>01OV018</th>\n",
       "      <th>01OV023</th>\n",
       "      <th>01OV026</th>\n",
       "      <th>01OV029</th>\n",
       "      <th>01OV030</th>\n",
       "      <th>01OV039</th>\n",
       "      <th>01OV041</th>\n",
       "      <th>01OV047</th>\n",
       "      <th>...</th>\n",
       "      <th>X20BR002</th>\n",
       "      <th>X20BR005</th>\n",
       "      <th>X20BR006</th>\n",
       "      <th>X20BR007</th>\n",
       "      <th>X20BR008</th>\n",
       "      <th>X21BR001</th>\n",
       "      <th>X21BR002</th>\n",
       "      <th>X21BR010</th>\n",
       "      <th>X22BR005</th>\n",
       "      <th>X22BR006</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.335938</td>\n",
       "      <td>-0.230469</td>\n",
       "      <td>0.1875</td>\n",
       "      <td>1.4375</td>\n",
       "      <td>-0.28125</td>\n",
       "      <td>-0.753906</td>\n",
       "      <td>0.101562</td>\n",
       "      <td>-1.117188</td>\n",
       "      <td>-0.140625</td>\n",
       "      <td>0.539062</td>\n",
       "      <td>...</td>\n",
       "      <td>1.171875</td>\n",
       "      <td>-1.125</td>\n",
       "      <td>0.882812</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.972656</td>\n",
       "      <td>-1.363281</td>\n",
       "      <td>-2.078125</td>\n",
       "      <td>0.457031</td>\n",
       "      <td>-0.207031</td>\n",
       "      <td>-0.199219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 1227 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       01OV007   01OV017  01OV018  01OV023  01OV026   01OV029   01OV030  \\\n",
       "A1BG  0.335938 -0.230469   0.1875   1.4375 -0.28125 -0.753906  0.101562   \n",
       "A1CF       NaN       NaN      NaN      NaN      NaN       NaN       NaN   \n",
       "\n",
       "       01OV039   01OV041   01OV047  ...  X20BR002  X20BR005  X20BR006  \\\n",
       "A1BG -1.117188 -0.140625  0.539062  ...  1.171875    -1.125  0.882812   \n",
       "A1CF       NaN       NaN       NaN  ...       NaN       NaN       NaN   \n",
       "\n",
       "      X20BR007  X20BR008  X21BR001  X21BR002  X21BR010  X22BR005  X22BR006  \n",
       "A1BG      -1.0  0.972656 -1.363281 -2.078125  0.457031 -0.207031 -0.199219  \n",
       "A1CF       NaN       NaN       NaN       NaN       NaN       NaN       NaN  \n",
       "\n",
       "[2 rows x 1227 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_proteomics = pd.read_parquet(file_cptac_proteomics)\n",
    "print(cptac_proteomics.shape)\n",
    "cptac_proteomics[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f6575f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12581, 1901)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X05BR044</th>\n",
       "      <th>X06BR006</th>\n",
       "      <th>X06BR005</th>\n",
       "      <th>X06BR003</th>\n",
       "      <th>X05BR045</th>\n",
       "      <th>X05BR016</th>\n",
       "      <th>X05BR043</th>\n",
       "      <th>X05BR042</th>\n",
       "      <th>X05BR038</th>\n",
       "      <th>X05BR029</th>\n",
       "      <th>...</th>\n",
       "      <th>C3L-00908</th>\n",
       "      <th>C3L-00907</th>\n",
       "      <th>C3L-00902</th>\n",
       "      <th>C3L-00817</th>\n",
       "      <th>C3L-00814</th>\n",
       "      <th>C3L-00813</th>\n",
       "      <th>C3L-00812</th>\n",
       "      <th>C3L-00800</th>\n",
       "      <th>C3L-00799</th>\n",
       "      <th>C3N-01808</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABCC9</th>\n",
       "      <td>0.335605</td>\n",
       "      <td>0.437866</td>\n",
       "      <td>0.289582</td>\n",
       "      <td>0.369700</td>\n",
       "      <td>0.274862</td>\n",
       "      <td>0.248999</td>\n",
       "      <td>0.300973</td>\n",
       "      <td>0.273792</td>\n",
       "      <td>0.365365</td>\n",
       "      <td>0.244641</td>\n",
       "      <td>...</td>\n",
       "      <td>0.106512</td>\n",
       "      <td>0.092456</td>\n",
       "      <td>0.147244</td>\n",
       "      <td>0.165088</td>\n",
       "      <td>0.112657</td>\n",
       "      <td>0.140601</td>\n",
       "      <td>0.172404</td>\n",
       "      <td>0.138284</td>\n",
       "      <td>0.102863</td>\n",
       "      <td>0.189004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGR3</th>\n",
       "      <td>0.060490</td>\n",
       "      <td>0.196469</td>\n",
       "      <td>0.279060</td>\n",
       "      <td>0.450851</td>\n",
       "      <td>0.028310</td>\n",
       "      <td>0.426830</td>\n",
       "      <td>0.044012</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.456943</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.010735</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005646</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007103</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 1901 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       X05BR044  X06BR006  X06BR005  X06BR003  X05BR045  X05BR016  X05BR043  \\\n",
       "ABCC9  0.335605  0.437866  0.289582  0.369700  0.274862  0.248999  0.300973   \n",
       "AGR3   0.060490  0.196469  0.279060  0.450851  0.028310  0.426830  0.044012   \n",
       "\n",
       "       X05BR042  X05BR038  X05BR029  ...  C3L-00908  C3L-00907  C3L-00902  \\\n",
       "ABCC9  0.273792  0.365365  0.244641  ...   0.106512   0.092456   0.147244   \n",
       "AGR3   0.000000  0.456943  0.000000  ...   0.002412   0.000000   0.010735   \n",
       "\n",
       "       C3L-00817  C3L-00814  C3L-00813  C3L-00812  C3L-00800  C3L-00799  \\\n",
       "ABCC9   0.165088   0.112657   0.140601   0.172404   0.138284   0.102863   \n",
       "AGR3    0.000000   0.000000   0.005646   0.000000   0.007103   0.000000   \n",
       "\n",
       "       C3N-01808  \n",
       "ABCC9   0.189004  \n",
       "AGR3    0.000000  \n",
       "\n",
       "[2 rows x 1901 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_transcriptomics = pd.read_parquet(file_cptac_transcriptomics)\n",
    "print(cptac_transcriptomics.shape)\n",
    "cptac_transcriptomics[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61155e1f",
   "metadata": {},
   "source": [
    "#### Process transcripts and proteins\n",
    "\n",
    "1. Handling Missing values:\n",
    "   - Proteins: Exclude proteins with >40% missing values \n",
    "   - Transcripts: Fill missing values with 0s\n",
    "\n",
    "2. Protein isoforms / Repeated samples: \n",
    "   - Aggregating by computing the mean                         \n",
    "                                                                               \n",
    "3. Consider only the transcripts with greater than 0 variance   \n",
    "4. Regress the cancer type from proteomics data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d1124d1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common proteins = 11058 and common samples = 1157\n"
     ]
    }
   ],
   "source": [
    "common_proteins = np.intersect1d(cptac_transcriptomics.index, cptac_proteomics.index)\n",
    "common_samples = np.intersect1d(cptac_transcriptomics.columns, cptac_proteomics.columns)\n",
    "print(\"Common proteins = \" + str(len(common_proteins)) + \" and common samples = \" + str(len(common_samples)))\n",
    "cptac_transcriptomics_subset = cptac_transcriptomics.reindex(common_samples, axis=1).T\n",
    "cptac_proteomics_subset = cptac_proteomics.reindex(common_samples, axis=1).reindex(common_proteins).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c66f35d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ABCC9</th>\n",
       "      <th>AGR3</th>\n",
       "      <th>TFF1</th>\n",
       "      <th>ABCC11</th>\n",
       "      <th>ABCC8</th>\n",
       "      <th>LMX1B</th>\n",
       "      <th>ABCA12</th>\n",
       "      <th>TMPRSS4</th>\n",
       "      <th>CEACAM5</th>\n",
       "      <th>PRAME</th>\n",
       "      <th>...</th>\n",
       "      <th>CNOT2</th>\n",
       "      <th>PPWD1</th>\n",
       "      <th>DDX24</th>\n",
       "      <th>IWS1</th>\n",
       "      <th>PDCL</th>\n",
       "      <th>CWC22</th>\n",
       "      <th>ZNF143</th>\n",
       "      <th>ZNF317</th>\n",
       "      <th>APOH</th>\n",
       "      <th>Study</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>0.438057</td>\n",
       "      <td>0.013974</td>\n",
       "      <td>0.009735</td>\n",
       "      <td>0.366004</td>\n",
       "      <td>0.013974</td>\n",
       "      <td>0.358992</td>\n",
       "      <td>0.256479</td>\n",
       "      <td>0.113873</td>\n",
       "      <td>0.447113</td>\n",
       "      <td>0.079922</td>\n",
       "      <td>...</td>\n",
       "      <td>0.650427</td>\n",
       "      <td>0.531460</td>\n",
       "      <td>0.704803</td>\n",
       "      <td>0.645089</td>\n",
       "      <td>0.535142</td>\n",
       "      <td>0.586527</td>\n",
       "      <td>0.521010</td>\n",
       "      <td>0.556072</td>\n",
       "      <td>0.013974</td>\n",
       "      <td>BrCa2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0D2</th>\n",
       "      <td>0.405471</td>\n",
       "      <td>0.237575</td>\n",
       "      <td>0.230412</td>\n",
       "      <td>0.073659</td>\n",
       "      <td>0.155145</td>\n",
       "      <td>0.510119</td>\n",
       "      <td>0.209696</td>\n",
       "      <td>0.091062</td>\n",
       "      <td>0.055237</td>\n",
       "      <td>0.505577</td>\n",
       "      <td>...</td>\n",
       "      <td>0.575599</td>\n",
       "      <td>0.527847</td>\n",
       "      <td>0.663623</td>\n",
       "      <td>0.592113</td>\n",
       "      <td>0.574797</td>\n",
       "      <td>0.540097</td>\n",
       "      <td>0.478587</td>\n",
       "      <td>0.571109</td>\n",
       "      <td>0.028262</td>\n",
       "      <td>BrCa2016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 12582 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ABCC9      AGR3      TFF1    ABCC11     ABCC8     LMX1B  \\\n",
       "index                                                                      \n",
       "TCGA-A2-A0CM  0.438057  0.013974  0.009735  0.366004  0.013974  0.358992   \n",
       "TCGA-A2-A0D2  0.405471  0.237575  0.230412  0.073659  0.155145  0.510119   \n",
       "\n",
       "                ABCA12   TMPRSS4   CEACAM5     PRAME  ...     CNOT2     PPWD1  \\\n",
       "index                                                 ...                       \n",
       "TCGA-A2-A0CM  0.256479  0.113873  0.447113  0.079922  ...  0.650427  0.531460   \n",
       "TCGA-A2-A0D2  0.209696  0.091062  0.055237  0.505577  ...  0.575599  0.527847   \n",
       "\n",
       "                 DDX24      IWS1      PDCL     CWC22    ZNF143    ZNF317  \\\n",
       "index                                                                      \n",
       "TCGA-A2-A0CM  0.704803  0.645089  0.535142  0.586527  0.521010  0.556072   \n",
       "TCGA-A2-A0D2  0.663623  0.592113  0.574797  0.540097  0.478587  0.571109   \n",
       "\n",
       "                  APOH     Study  \n",
       "index                             \n",
       "TCGA-A2-A0CM  0.013974  BrCa2016  \n",
       "TCGA-A2-A0D2  0.028262  BrCa2016  \n",
       "\n",
       "[2 rows x 12582 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_transcriptomics_study = cptac_transcriptomics_subset.merge(cptac_sample, right_index=True,\n",
    "                                                               left_index=True).reset_index().sort_values(['Study', 'index']).set_index('index')\n",
    "cptac_transcriptomics_study[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a76a608",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>Study</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>-0.640625</td>\n",
       "      <td>7.234375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.305664</td>\n",
       "      <td>-3.3125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.417969</td>\n",
       "      <td>-0.328125</td>\n",
       "      <td>-1.158203</td>\n",
       "      <td>...</td>\n",
       "      <td>1.058594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.683594</td>\n",
       "      <td>-0.330078</td>\n",
       "      <td>-0.344727</td>\n",
       "      <td>-0.015625</td>\n",
       "      <td>BrCa2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0D2</th>\n",
       "      <td>-0.281250</td>\n",
       "      <td>-0.703125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.789062</td>\n",
       "      <td>-1.1875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.832031</td>\n",
       "      <td>-0.367188</td>\n",
       "      <td>-1.148438</td>\n",
       "      <td>...</td>\n",
       "      <td>2.484375</td>\n",
       "      <td>1.9375</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.290992</td>\n",
       "      <td>0.567383</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>-0.657227</td>\n",
       "      <td>0.289062</td>\n",
       "      <td>BrCa2016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 11059 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   A2M     A2ML1  A4GALT  A4GNT      AAAS    AACS  AADAT  \\\n",
       "index                                                                      \n",
       "TCGA-A2-A0CM -0.640625  7.234375     NaN    NaN  0.305664 -3.3125    NaN   \n",
       "TCGA-A2-A0D2 -0.281250 -0.703125     NaN    NaN  0.789062 -1.1875    NaN   \n",
       "\n",
       "                 AAGAB      AAK1      AAMP  ...    ZWILCH   ZWINT  ZXDA  ZXDC  \\\n",
       "index                                       ...                                 \n",
       "TCGA-A2-A0CM -1.417969 -0.328125 -1.158203  ...  1.058594     NaN   NaN   NaN   \n",
       "TCGA-A2-A0D2 -0.832031 -0.367188 -1.148438  ...  2.484375  1.9375   NaN   NaN   \n",
       "\n",
       "                ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3     Study  \n",
       "index                                                                     \n",
       "TCGA-A2-A0CM       NaN -0.683594 -0.330078 -0.344727 -0.015625  BrCa2016  \n",
       "TCGA-A2-A0D2 -1.290992  0.567383  0.875000 -0.657227  0.289062  BrCa2016  \n",
       "\n",
       "[2 rows x 11059 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac_proteomics_study = cptac_proteomics_subset.merge(cptac_sample, right_index=True,\n",
    "                                                       left_index=True).reset_index().sort_values(['Study', \n",
    "                                                                                                   'index']).set_index('index')\n",
    "cptac_proteomics_study[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "42454d38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Study\n",
       "Pdac        140\n",
       "BrCa2020    122\n",
       "LUAD        110\n",
       "ccRCC       110\n",
       "HNSCC       109\n",
       "LSCC        108\n",
       "OvCa2016    105\n",
       "GBM          99\n",
       "UCEC         95\n",
       "OvCa2020     82\n",
       "BrCa2016     77\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_count_per_study = cptac_proteomics_subset.merge(cptac_sample, right_index=True, left_index=True)['Study'].value_counts()\n",
    "sample_count_per_study.sort_values(ascending=False, inplace=True)\n",
    "sample_count_per_study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "85146ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_prediction(protein, x_train, y_train, x_test, y_test):    \n",
    "    # check if predictor df is null - np.nan is an instance of float\n",
    "    if(isinstance(x_train, float) | isinstance(x_test, float)):\n",
    "        return(pd.Series(index=y_test.index, name=protein, dtype=np.float64))\n",
    "    \n",
    "    common_samples = np.intersect1d(x_train.dropna().index, y_train.dropna().index)\n",
    "    x_train_subset = x_train.reindex(common_samples)\n",
    "    y_train_subset = y_train.reindex(common_samples)\n",
    "    \n",
    "    common_samples = np.intersect1d(x_test.dropna().index, y_test.dropna().index)\n",
    "    x_test_subset = x_test.reindex(common_samples)\n",
    "    y_test_subset = y_test.reindex(common_samples)\n",
    "        \n",
    "    if(len(x_train_subset) < 2):\n",
    "        return(pd.Series(index=y_test_subset.index, name=protein, dtype=np.float64))\n",
    "    if(isinstance(x_train_subset, pd.Series)):\n",
    "        x_train_subset = x_train_subset.values.reshape(-1,1)\n",
    "        x_test_subset = x_test_subset.values.reshape(-1,1)\n",
    "\n",
    "    predictor = LinearRegression()\n",
    "    predictor.fit(x_train_subset, y_train_subset)\n",
    "    \n",
    "    y_test_pred = pd.Series(predictor.predict(x_test_subset), index=y_test_subset.index, name=protein)\n",
    "                        \n",
    "    return(y_test_pred)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb3c676",
   "metadata": {},
   "source": [
    "### Scenarios: \n",
    "\n",
    "1. self-mRNA\n",
    "2. self-mRNA + CORUM\n",
    "3. self-mRNA + STRING 400\n",
    "4. self-mRNA + STRING 800\n",
    "5. self-mRNA + STRING physical interactions 400\n",
    "6. self-mRNA + STRING physical interactions 800\n",
    "7. self-mRNA + VAE\n",
    "8. self-mRNA + VAE + CORUM \n",
    "9. self-mRNA + VAE + STRING 400\n",
    "10. self-mRNA + VAE + STRING 800\n",
    "11. self-mRNA + VAE + STRING physical interactions 400\n",
    "12. self-mRNA + VAE + STRING physical interactions 800"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dbfd9a22",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_studywise_data(study):\n",
    "    # ensuring the proteomics and transcriptomics data is available for at least 60% of the samples\n",
    "    proteomics = cptac_proteomics_study[cptac_proteomics_study['Study'] == study].drop(columns=['Study'])\n",
    "    proteomics.dropna(thresh=0.6*(len(proteomics)), axis=1, inplace=True)    \n",
    "    transcriptomics = cptac_transcriptomics_study[cptac_transcriptomics_study['Study'] == study].drop(columns=['Study']) \n",
    "    # Filling the data with less than 60% null values with 0\n",
    "    transcriptomics = transcriptomics.dropna(thresh=0.6*(len(transcriptomics)), axis=1).fillna(0)\n",
    "    \n",
    "    # Select only those proteins with both proteomic and transcriptomic data available for at least 60% samples\n",
    "    common_proteins = np.intersect1d(transcriptomics.columns, proteomics.columns)\n",
    "    transcriptomics = transcriptomics.reindex(common_proteins, axis=1)\n",
    "    proteomics = proteomics.reindex(common_proteins, axis=1)\n",
    "    \n",
    "    return(transcriptomics, proteomics)\n",
    "  \n",
    "def run_parallelized_baseline_pipeline(study, feature_set, desired_filename): \n",
    "    X, y = get_studywise_data(study)\n",
    "    print(\"Number of proteins considered = \", len(y.columns))\n",
    "    \n",
    "    all_folds_results = []\n",
    "    kf = KFold(n_splits=5, random_state=None, shuffle=False)\n",
    "    for fold, (train, test) in enumerate(kf.split(X)):\n",
    "        X_train = X.iloc[train]\n",
    "        X_test = X.iloc[test]\n",
    "        y_train = y.reindex(X_train.index)\n",
    "        y_test = y.reindex(X_test.index)\n",
    "        all_folds_results.append(y.columns.to_series().apply(lambda p: \\\n",
    "                                      perform_prediction(p, X_train[p], y_train[p], X_test[p], y_test[p])))\n",
    "    pd.concat(all_folds_results, axis=1).to_parquet(get_output_path(desired_filename))\n",
    "    print(\"Completed \" + study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad98918c",
   "metadata": {},
   "outputs": [],
   "source": [
    "parquet_extension = \".parquet\"\n",
    "BASELINE = 'Baseline'\n",
    "baselinefile_prefix = \"baseline_LR_selfmRNA_\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bd2e42fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Baseline\n",
      "Number of proteins considered =  6433\n",
      "Completed Pdac\n",
      "Number of proteins considered =  7421\n",
      "Completed BrCa2020\n",
      "Number of proteins considered =  7310\n",
      "Completed LUAD\n",
      "Number of proteins considered =  6413\n",
      "Completed ccRCC\n",
      "Number of proteins considered =  7113\n",
      "Completed HNSCC\n",
      "Number of proteins considered =  7905\n",
      "Completed LSCC\n"
     ]
    }
   ],
   "source": [
    "print(BASELINE)\n",
    "for index, value in sample_count_per_study.head(6).items():\n",
    "    desired_filename = baselinefile_prefix + index + parquet_extension\n",
    "    run_parallelized_baseline_pipeline(index, BASELINE, desired_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1497c7f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_cptac_proteomics = get_input_path('CPTAC_proteomics_linkedomics.parquet')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
